alns/select/OperatorSelection.py

```diff
@@ -0,0 +1,111 @@
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+import numpy as np
+from numpy.random import RandomState
+
+from alns.State import State
+
+
+class OperatorSelection(ABC):
+    """
+    Base class from which to implement an operator selection scheme.
+
+    Parameters
+    ----------
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional keyword argument. Matrix that indicates coupling between
+        destroy and repair operators. Entry (i, j) is 1 if destroy operator i
+        can be used in conjunction with repair operator j and 0 otherwise.
+    """
+
+    def __init__(
+        self,
+        num_destroy: int,
+        num_repair: int,
+        *,
+        op_coupling: Optional[np.ndarray] = None,
+    ):
+        self._validate_arguments(num_destroy, num_repair, op_coupling)
+
+        self._num_destroy = num_destroy
+        self._num_repair = num_repair
+        self._op_coupling = (
+            op_coupling
+            if op_coupling is not None
+            else np.ones((num_destroy, num_repair))
+        )
+
+    @property
+    def num_destroy(self) -> int:
+        return self._num_destroy
+
+    @property
+    def num_repair(self) -> int:
+        return self._num_repair
+
+    @property
+    def op_coupling(self) -> np.ndarray:
+        return self._op_coupling
+
+    @abstractmethod
+    def __call__(
+        self, rnd: RandomState, best: State, curr: State
+    ) -> Tuple[int, int]:
+        """
+        Determine which destroy and repair operator pair to apply in this
+        iteration.
+
+        Parameters
+        ----------
+        rnd_state
+            Random state object, to be used for random number generation.
+        best
+            The best solution state observed so far.
+        current
+            The current solution state.
+
+        Returns
+        -------
+        A tuple of (d_idx, r_idx), which are indices into the destroy and
+        repair operator lists, respectively.
+        """
+        return NotImplemented
+
+    @abstractmethod
+    def update(self, candidate: State, d_idx: int, r_idx: int, outcome: int):
```

Comment:
Suggestion: maybe also introduce the outcome enum already, now that we are changing the name of `s_idx` to `outcome`?

Reply:
It's probably easier if I do this when implementing #94. Also because #95 has a few docstring changes, so it's easier to merge if I don't change the outcomes yet. I have renamed `outcome` to `s_idx` as well so I can do it all at once!